### PR TITLE
Add environment parameter in config

### LINKED ISSUE #175: (feature) Add environment parameters in the config
# Requirements
Add an environment parameter in the config.
This environment field should be added to all users in the param so we can add rules based on the environment and have the same config file for all your environment.

### PR DESCRIPTION
**Pull request by mtwstudios**
# Description
Checking the current environment / flagging by environment is a common use cases for projects I'm involved in, so this adds environment awareness to feature flags. You can set environment as a field in the Config passed to Init, then check them in rule (e.g. `rule: env == "dev"`)

I realize I need to address tests/docs, but wanted to get feedback on the approach first to see if it was acceptable.

# Changes include
- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking changes (change that is not backward-compatible and/or changes current functionality)

# Closes issue(s)

Resolve #175

# Checklist
- [x] I have tested this code
- [x] I have added unit test to cover this code
- [x] I have updated the documentation (README.md and /docs)
- [x] I have followed the [contributing guide](CONTRIBUTING.md)

### CODE REVIEW

File: internal/flagv1/flag_data.go [hunk 1]
```diff
@@ -77,9 +78,8 @@ func (f *FlagData) Value(flagName string, user ffuser.User) (interface{}, string
 
 func (f *FlagData) isExperimentationOver() bool {
 	now := time.Now()
-	return f.Rollout != nil && f.Rollout.Experimentation != nil && (
-		(f.Rollout.Experimentation.Start != nil && now.Before(*f.Rollout.Experimentation.Start)) ||
-			(f.Rollout.Experimentation.End != nil && now.After(*f.Rollout.Experimentation.End)))
+	return f.Rollout != nil && f.Rollout.Experimentation != nil && ((f.Rollout.Experimentation.Start != nil && now.Before(*f.Rollout.Experimentation.Start)) ||
```
Comment (thomaspoignant) [inline internal/flagv1/flag_data.go:81]: can we keep the same indentation as before to avoid having a too-long line?

File: internal/flagv1/flag_data.go [hunk 2]
```diff
@@ -2,13 +2,14 @@ package flagv1
...
-	"github.com/thomaspoignant/go-feature-flag/ffuser"
-	"github.com/thomaspoignant/go-feature-flag/internal/utils"
 	"math"
 	"strconv"
 	"strings"
 	"time"
+
+	"github.com/nikunjy/rules/parser"
+	"github.com/thomaspoignant/go-feature-flag/ffuser"
+	"github.com/thomaspoignant/go-feature-flag/internal/utils"
```
Comment (thomaspoignant) [inline internal/flagv1/flag_data.go:12]: Can we keep the import as before, there is no change in them just a different order.

File: internal/flagv1/flag_data.go [hunk 3]
```diff
@@ -113,7 +113,9 @@ func (f *FlagData) evaluateRule(user ffuser.User) bool {
 	}
 
 	// Evaluate the rule on the user.
-	return parser.Evaluate(f.getRule(), utils.UserToMap(user))
+	userMap := utils.UserToMap(user)
+	userMap["env"] = environment
```
Comment (thomaspoignant) [inline internal/flagv1/flag_data.go:117]: Since by default `env` is **empty** I will avoid having the key `env` in the map if there is no value.
```suggestion
	if environment != "" {
		userMap["env"] = environment
	}	
```

What do you think about that?

File: config.go [hunk 4]
```diff
@@ -3,10 +3,11 @@ package ffclient
 import (
 	"context"
 	"errors"
-	"github.com/thomaspoignant/go-feature-flag/ffnotifier"
 	"log"
 	"time"
 
+	"github.com/thomaspoignant/go-feature-flag/ffnotifier"
```
Comment (thomaspoignant) [inline config.go:9]: Can we keep the import as before, there is no change in them just a different order.

File: internal/flagv1/flag_priv_test.go [hunk 5]
```diff
@@ -1,10 +1,11 @@
 package flagv1
 
 import (
-	"github.com/stretchr/testify/assert"
 	"testing"
 	"time"
 
+	"github.com/stretchr/testify/assert"
```
Comment (thomaspoignant) [inline internal/flagv1/flag_priv_test.go:7]: Can we keep the import as before, there is no change in them just a different order.

File: internal/flagv1/flag_pub_test.go [hunk 6]
```diff
@@ -2,11 +2,12 @@ package flagv1_test
 
 import (
 	"fmt"
+	"testing"
```
Comment (thomaspoignant) [inline internal/flagv1/flag_pub_test.go:5]: Can we keep the import as before, there is no change in them just a different order.

File: variation.go [hunk 7]
```diff
@@ -2,6 +2,7 @@ package ffclient
 
 import (
 	"fmt"
+
```
Comment (thomaspoignant) [inline variation.go:5]: Can we remove this blank line since it is not related to your change

**Review by thomaspoignant — CHANGES_REQUESTED** [review]
Hey, @mtwstudios thanks for your contribution this is awesome.
The approach you have is the one I had in mind when writing the issue so it looks good to me.

Despite that I made some minor changes request in your pull request to avoid changing unused things.

I would also suggest adding some tests where you are testing the environment feature because in your test you are always sending an empty `environment` so it will not test that this is working as expected.

File: internal/flagv1/flag_data.go (hunk 3 above)
Comment (mtwstudios) [reply internal/flagv1/flag_data.go:117]: makes sense to me!

**Comment by mtwstudios** [comment]
Thanks for the feedback @thomaspoignant ! I think I've addressed everything, including the cleanup, and adding tests and docs for the change. Let me know if there's any other changes you think are needed.

**Comment by mtwstudios** [comment]
Hey @thomaspoignant, yep - with this solution, env would be specified in the rule.

The advantage of having it be specified in the rule is that it allows for more flexibility for various use cases. In particular:

- **Enabling for multiple (but not all) environments.** This is a very common use case on teams I've been on - for example to enable a flag in local development environments and staging environments but not in production
  - e.g. `rule: (env == "dev") or (env == "staging")`
 - **Enabling the flag based on certain rules only in certain environments.** For example it's enabled for all users in non-production environments, but in production it's only enabled for specific users
   - e.g. `rule: (env != "prod") or (user_id == 1234)`

I think having `environment` separate from `rule`, while might be easier in the simplest case, prevents it from being used in cases like the examples above. This would make it much less useful/powerful in my opinion.

> My only concern is that we cannot have 2 different configurations for the same flag, and it could be a limitation.

I'm not sure what you mean by this - can you give an example?

**Review by thomaspoignant — APPROVED** [review]
